tests/integrations/raftstore/test_joint_consensus.rs

```diff
@@ -0,0 +1,475 @@
+    let rx = cluster
+        .async_request(put_request(&region, 1, b"k3", b"v3"))
+        .unwrap();
+    assert_eq!(
+        rx.recv_timeout(Duration::from_millis(100)),
+        Err(mpsc::RecvTimeoutError::Timeout)
+    );
+    cluster.clear_send_filters();
+
+    // Isolated peer 3, so the new configuation can't reach quorum
+    cluster.add_send_filter(IsolationFilterFactory::new(3));
+    let rx = cluster
+        .async_request(put_request(&region, 1, b"k4", b"v4"))
+        .unwrap();
+    assert_eq!(
+        rx.recv_timeout(Duration::from_millis(100)),
+        Err(mpsc::RecvTimeoutError::Timeout)
+    );
+    cluster.clear_send_filters();
+
+    // Leave joint
+    pd_client.must_leave_joint(region_id);
+
+    // Isolated peer 2, but it is not in quorum any more
+    cluster.add_send_filter(IsolationFilterFactory::new(2));
+    cluster.must_put(b"k5", b"v5");
+    must_get_equal(&cluster.get_engine(3), b"k5", b"v5");
+}
+
+/// test_replace_peer_joint_state testing that when replace peer, request can
+/// be handled as usual, even two nodes crashed and leader can be replaced by
+/// in joint state
+#[test]
+fn test_joint_replace_peers() {
+    let mut cluster = new_node_cluster(0, 4);
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+    let region_id = cluster.run_conf_change();
+    cluster.must_transfer_leader(region_id, new_peer(1, 1));
+
+    cluster.must_put(b"k1", b"v1");
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(3), b"k1", b"v1");
+
+    // Replace peers
+    // Enter joint, now we have C_old(1, 2, 3) and C_new(1, 2, 4)
+    pd_client.must_joint_confchange(
+        region_id,
+        vec![
+            (ConfChangeType::AddLearnerNode, new_learner_peer(3, 3)),
+            (ConfChangeType::AddNode, new_peer(4, 4)),
+        ],
+    );
+    must_get_equal(&cluster.get_engine(4), b"k1", b"v1");
+
+    // Isolated node 3 and node 4
+    cluster.add_send_filter(IsolationFilterFactory::new(3));
+    cluster.add_send_filter(IsolationFilterFactory::new(4));
+
+    // Request can be handle as usual
+    cluster.must_put(b"k2", b"v2");
+    must_get_equal(&cluster.get_engine(2), b"k2", b"v2");
+    must_get_none(&cluster.get_engine(3), b"k2");
+    must_get_none(&cluster.get_engine(4), b"k2");
+
+    // Leave joint
+    pd_client.must_leave_joint(region_id);
+
+    cluster.clear_send_filters();
+    must_get_equal(&cluster.get_engine(3), b"k2", b"v2");
+    must_get_equal(&cluster.get_engine(4), b"k2", b"v2");
+
+    // Replace leader
+    // Enter joint, now we have C_old(1, 2, 4) and C_new(2, 3, 4)
+    pd_client.must_joint_confchange(
+        region_id,
+        vec![
+            (ConfChangeType::AddLearnerNode, new_learner_peer(1, 1)),
+            (ConfChangeType::AddNode, new_peer(3, 3)),
```

Comment:
Actually adding back node as voter is undefined, not sure if it works generally.

Reply:
I add a new node instead, also I change some checks of demoting voter and come along with some test cases, PTAL.